### PR TITLE
Issue #293

### PR DESCRIPTION
**Pull request by mitjap**
Mocking location is working but is not perfect. 
-Right now everyone can mock location. How do we know if user is authorized for such action.
-Translations are still not working. Any idea what I did wrong?
#293

### CODE REVIEW

**Comment by mitar** [comment]
> Right now everyone can mock location. How do we know if user is authorized for such action.

As I described, only staff users should be allowed to do that. Check definition of user document, I think you will find an answer there.

**Comment by mitar** [comment]
> Translations are still not working. Any idea what I did wrong?

Have you recompiled them? Read our documentation how to work with translations.

File: piplmesh/frontend/context_processors.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 from django.utils import translation
 
 from piplmesh import urls
+from piplmesh import nodes
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: No need for two lines. You can merge these two imports into one line. Order alphabetically.

File: piplmesh/frontend/middleware.py [hunk 2]
```diff
@@ -8,6 +8,7 @@ class NodesMiddleware(object):
     def process_request(self, request):
         outside_url = urlresolvers.reverse('outside')
         request.node = nodes.get_node(request)
+        request.nodes = nodes.get_all_nodes()
```
Comment (mitar) [inline piplmesh/frontend/middleware.py]: No, we do not want that EVERY request has to load all nodes. Staff users will be a big minority and why would you load huge data for everybody?

Also, middleware is not the right place to store data you want to use in templates. Please read about context processors.

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 3]
```diff
@@ -181,6 +181,12 @@ input.gsc-search-button:hover,
     height: 50px;
 }
 
+#header #mock_location {
+    position:absolute;
+    top: 5px;
+    right:300px;
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: Space after `:`.

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 4]
```diff
@@ -181,6 +181,12 @@ input.gsc-search-button:hover,
     height: 50px;
 }
 
+#header #mock_location {
+    position:absolute;
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: Ehm, absolute? Why? It should simply be another element in the header menu (next to login, logout, and other entries).

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 5]
```diff
@@ -181,6 +181,12 @@ input.gsc-search-button:hover,
     height: 50px;
 }
 
+#header #mock_location {
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: HTML element IDs have to be unique over all page. There is no reason to say that uniquely identified element should be child of another niquely identified element. You have already uniquely identified it in the first place. :-)

File: piplmesh/frontend/templates/header.html [hunk 6]
```diff
@@ -21,6 +21,18 @@
                 {% trans "See your query results on " %}<a id="google_results_link">Google</a>.
             {% endif %}
         </div>
+        <div id="mock_location">
+            <form method="post" action="{% url "mock_location" %}" id="mock_location_form">
+                {% csrf_token %}
+                {% trans "Mock location" %}:
+                <select name="mocked_location" onchange="this.form.submit();">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: No inline JavaScript. Use jQuery to attach event handler to elements.

File: piplmesh/frontend/templates/header.html [hunk 7]
```diff
@@ -21,6 +21,18 @@
                 {% trans "See your query results on " %}<a id="google_results_link">Google</a>.
             {% endif %}
         </div>
+        <div id="mock_location">
+            <form method="post" action="{% url "mock_location" %}" id="mock_location_form">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Don't construct forms manually. Use Django forms for this. (Also that you learn how to use it.)

File: piplmesh/frontend/views.py [hunk 8]
```diff
@@ -147,3 +148,15 @@ def panels_order(request):
         number_of_columns = request.GET['number_of_columns']
         panels = request.user.panels_order.get(number_of_columns, [])
         return http.HttpResponse(simplejson.dumps(panels), mimetype='application/json')
+
+def mock_location(request):
+    next = request.REQUEST.get('next', None)
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Where do you set next? Why it would be configured? This is your job to set it.

Comment (mitar) [inline piplmesh/frontend/views.py]: And why mixing GET and POST? Why not have next URL as part of POST?

File: piplmesh/frontend/views.py [hunk 9]
```diff
@@ -147,3 +148,15 @@ def panels_order(request):
...
+    if not next:
+        next = request.META.get('HTTP_REFERER', None)
```
Comment (mitar) [inline piplmesh/frontend/views.py]: We don't want to depend on browser features. If we want user to be redirected back where she came from, we should make sure we always have this data. And then it is not necessary to check browser headers for this.

File: piplmesh/frontend/views.py [hunk 10]
```diff
@@ -147,3 +148,15 @@ def panels_order(request):
...
+    if not next:
+        next = '/'
+    response = http.HttpResponseRedirect(next)
+    if request.method == 'POST':
+        mock_location = request.POST.get('mocked_location', None)
```
Comment (mitar) [inline piplmesh/frontend/views.py]: You are not doing any verification here on the value POST. This is bad.

You should really be using Django forms.

File: piplmesh/frontend/views.py [hunk 11]
```diff
@@ -147,3 +148,15 @@ def panels_order(request):
         number_of_columns = request.GET['number_of_columns']
         panels = request.user.panels_order.get(number_of_columns, [])
         return http.HttpResponse(simplejson.dumps(panels), mimetype='application/json')
+
+def mock_location(request):
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Use class-based views. Not function views.

File: piplmesh/nodes/__init__.py [hunk 12]
```diff
@@ -63,10 +64,22 @@ def get_node(request):
 
     node = None
     try:
-        node_id = request.session[SESSION_KEY]
-        backend_path = request.session[BACKEND_SESSION_KEY]
-        backend = load_backend(backend_path)
-        node = backend.get_node(node_id)
+        mocked_node_id = request.session[MOCKED_SESSION_KEY]
```
Comment (mitar) [inline piplmesh/nodes/__init__.py]: This code is similar to what I have in mind, but needs more improvement:
- you want this code to be fast as it is called on every request, iterating over all backends in not a good idea, this is why we store backend in the session
- so in the view above, you should mingle with session, not here, so in the view you trick the system to believe it is on that location you are mocking, and then leave this code here to blindly trust it
- so I would just override stored nodes_id and backend to the chosen one, I would only have a session value which tells that mocking is in effect so that you can provide user feedback to the staff user (showing that mocking is in effect and which node it is mocked to (the one in `request.node`))

File: piplmesh/nodes/__init__.py [hunk 13]
```diff
@@ -7,6 +7,7 @@
 LONGITUDE_SESSION_KEY = '_longitude'
 
 SESSION_KEY = '_nodes_id'
+MOCKED_SESSION_KEY = '_nodes_id_mocked'
```
Comment (mitar) [inline piplmesh/nodes/__init__.py]: If you are adding new session keys, you should extend also `flush_session` function, and also check elsewhere if there is something similar to add too.

**Comment by mitar** [comment]
Please update from main repository. It cannot be automatically merged.

File: piplmesh/frontend/forms.py [hunk 14]
```diff
@@ -10,3 +12,8 @@ class ContactForm(forms.Form):
     subject = forms.CharField(label=_("Subject"))
     email = forms.EmailField(label=_("Your e-mail address"))
     message = forms.CharField(widget=widgets.Textarea())
+
+class LocationsForm(forms.Form):
+    locationChoices = [("%s_%s" % (backend.get_full_name(), node.id), node.name) for backend, node in nodes.get_all_nodes_with_backends()]
+    locationChoices.insert(0, ('-1', 'None'))
+    locations = forms.ChoiceField(choices=locationChoices, label=None, required=False)
```
Comment (mitjap) [inline piplmesh/frontend/forms.py]: Initial value should be set here, but I'm not sure how to create constructor with arguments (so that request can be passed). What am I doing wrong:

```
def __init__(self, request):
    locationChoices = [("%s_%s" % (backend.get_full_name(), node.id), node.name) for backend, node in nodes.get_all_nodes_with_backends()]
    locationChoices.insert(0, ('-1', 'None'))
    init = get_initial_value_from_request
    locations = forms.ChoiceField(choices=locationChoices, initial=init, label=None, required=False)
    forms.Form.__init__(self)
```

File: piplmesh/frontend/context_processors.py [hunk 15]
```diff
@@ -21,4 +20,7 @@ def global_vars(request):
         # Variables
         'logo_url': "piplmesh/images/logo-%s.png" % translation.get_language(),
         'request_get_next': request.REQUEST.get(auth.REDIRECT_FIELD_NAME),
+
+        # Forms
+        'locations_form' : forms.LocationsForm() if request.user and request.user.is_staff else None
```
Comment (mitjap) [inline piplmesh/frontend/context_processors.py]: request would be passed here (see comment below)

File: piplmesh/frontend/static/piplmesh/js/common.js [hunk 16]
```diff
@@ -22,6 +22,11 @@ $(document).ready(function () {
         }
     );
 
+    // Locations
+    $('#id_locations').change(function (event) {
+        $(this).closest("form").submit();
```
Comment (mitjap) [inline piplmesh/frontend/static/piplmesh/js/common.js]: Selecting an item should submit parent form. Is this the right way to do it?

Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common.js]: Yes.

Use `'` for constant strings.

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 17]
```diff
@@ -181,6 +181,10 @@ input.gsc-search-button:hover,
     height: 50px;
 }
 
+#mock_location {
+    float: right;
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: I don't like floating here. It should be simple addition to the menu list in the header.

File: piplmesh/frontend/forms.py (hunk 14 above)
Comment (mitar) [inline piplmesh/frontend/forms.py]: I provided you already with an [example](https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/account/forms.py#L60).

**Comment by mitar** [comment]
Pull requests is failing tests. See above (Travis build). Please research why.

**Comment by mitjap** [comment]
Tests are now working. Commit speaks for itself.

**Comment by mitar** [comment]
This pull request cannot be automatically merged. Please update from main repository.

File: piplmesh/frontend/context_processors.py (hunk 15 above)
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: Not really correct check. Much better check would be to check if user is authenticated.

And also add comma to the end.

```
'locations_form' : forms.LocationsForm() if getattr(request, 'user', None) and request.user.is_authenticated() and request.user.is_staff else None,
```

File: piplmesh/frontend/forms.py [hunk 18]
```diff
@@ -1,6 +1,8 @@
+
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: Unnecessary empty line.

File: piplmesh/frontend/forms.py [hunk 19]
```diff
@@ -1,6 +1,8 @@
...
 from django import forms
 from django.forms import widgets
 from django.utils.translation import ugettext_lazy as _
+from piplmesh import nodes
```
Comment (mitar) [inline piplmesh/frontend/forms.py:5]: You should group imports. So this is another group, different than Django imports. So newline before it.

File: piplmesh/frontend/static/piplmesh/js/common.js [hunk 20]
```diff
@@ -22,6 +22,11 @@ $(document).ready(function () {
         }
     );
 
+    // Locations
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common.js]: If we would be really strict, this code should be provided to users which are not staff.

Probably we have to define another JavaScript file which is included only for those users. We might later on even have more stuff there.

Comment (mitjap) [inline piplmesh/frontend/static/piplmesh/js/common.js]: (you have a typo in your comment) 

Never thought about that :) This solution would be good.

File: piplmesh/nodes/__init__.py [hunk 21]
```diff
@@ -67,6 +68,10 @@ def get_node(request):
         backend_path = request.session[BACKEND_SESSION_KEY]
         backend = load_backend(backend_path)
         node = backend.get_node(node_id)
+
+        if allow_mocking and request.session[MOCKING_SESSION_KEY]:
```
Comment (mitar) [inline piplmesh/nodes/__init__.py]: You probably want here `request.session.get(MOCKING_SESSION_KEY, False)`.

File: piplmesh/frontend/static/piplmesh/js/common.js (hunk 20 above)
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common.js]: Then do it. :-)

Comment (mitjap) [inline piplmesh/frontend/static/piplmesh/js/common.js]: In another ticket or this one?

Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common.js]: What do you mean? On what? I am commenting what all you have yet to fix with this pull request.

File: piplmesh/nodes/__init__.py (hunk 21 above)
Comment (mitjap) [inline piplmesh/nodes/__init__.py]: In case key is not present exception will happen and will not execute that statement. I'll fix it anyway.

File: piplmesh/frontend/templates/header.html [hunk 22]
```diff
@@ -21,8 +21,15 @@
                 {% trans "See your query results on " %}<a id="google_results_link">Google</a>.
             {% endif %}
         </div>
+
         {% with next|default:request_get_next|default:request.get_full_path as next_url %}
             <ul class="user">
+                {% with locations_form as form %}
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: This should be displayed only to authenticated and staff users, no?

File: piplmesh/frontend/static/piplmesh/js/home.js [hunk 23]
```diff
@@ -250,13 +250,13 @@ function addComment(comment) {
     var post_url = $('.post').first().data('post').resource_uri;
 
     $.ajax({
-        type: 'POST',
+        'type': 'POST',
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/home.js:264]: This all is completely unrelated stuff and would probably have to be another pull request. But leave it now.

File: piplmesh/frontend/views.py [hunk 24]
```diff
@@ -64,6 +65,23 @@ class UserView(detail.DetailView):
     slug_field = 'username'
     slug_url_kwarg = 'username'
 
+class LocationsView(generic_views.FormView):
+    form_class = forms.LocationsForm
+    success_url = urlresolvers.reverse_lazy('home')
```
Comment (mitar) [inline piplmesh/frontend/views.py:72]: Not good. It should redirect back to initial page. This is not necessary home. You can add TODO here.

File: piplmesh/frontend/templates/header.html (hunk 22 above)
Comment (mitjap) [inline piplmesh/frontend/templates/header.html]: It does. If users does not have permission 'locations_form' will be undefined (None) and will not include that code. I don't like to duplicate same conditions. Its usually prone to errors.

File: piplmesh/frontend/views.py (hunk 24 above)
Comment (mitar) [inline piplmesh/frontend/views.py:72]: initial page -> initiator page.

File: piplmesh/frontend/templates/header.html (hunk 22 above)
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: It will still display `<li></li>`. No?

Comment (mitjap) [inline piplmesh/frontend/templates/header.html]: I'll check. If it doesn't, do I leave it like it is?

Comment (mitar) [inline piplmesh/frontend/templates/header.html]: So there should not be any output here. If there is any output, you should put if around. If there is none, then it is good.

File: piplmesh/frontend/views.py [hunk 25]
```diff
@@ -64,6 +65,23 @@ class UserView(detail.DetailView):
...
+
+    def form_valid(self, form):
+        location = form.cleaned_data['locations']
+
+        if location == '-1':
+            nodes.flush_session(self.request)
+        else:
+            node_backend, node_id = location.rsplit('_')
+            self.request.session[nodes.SESSION_KEY] = int(node_id)
```
Comment (mitar) [inline piplmesh/frontend/views.py]: No, `node_id` is not necessary int. We do not define this anywhere in the API. It can be arbitrary string defined by the backend. So you should not convert it here.

File: piplmesh/frontend/views.py [hunk 26]
```diff
@@ -64,6 +65,23 @@ class UserView(detail.DetailView):
...
+    form_class = forms.LocationsForm
+    success_url = urlresolvers.reverse_lazy('home')
+
+    def form_valid(self, form):
+        location = form.cleaned_data['locations']
+
+        if location == '-1':
+            nodes.flush_session(self.request)
+        else:
+            node_backend, node_id = location.rsplit('_')
```
Comment (mitar) [inline piplmesh/frontend/views.py]: As explained bellow, `node_id` could in theory contain `_`. I would suggest that you use `-` instead and search from left. Because Python modules and classes cannot contain `-` in their names. So `name_backend` cannot contain `-`.

**Comment by mitar** [comment]
As I said, pull request cannot be merged. Please update from main repository.

**Comment by mitar** [comment]
There is user feedback missing that mocking is in progress and to which node it is locked. You have to also do that. (Or open another ticket.) Probably you just have to provide initial data to the form.

File: piplmesh/frontend/forms.py [hunk 27]
```diff
@@ -10,3 +12,8 @@ class ContactForm(forms.Form):
     subject = forms.CharField(label=_("Subject"))
     email = forms.EmailField(label=_("Your e-mail address"))
     message = forms.CharField(widget=widgets.Textarea())
+
+class LocationsForm(forms.Form):
+    locationChoices = [("%s_%s" % (backend.get_full_name(), node.id), node.name) for backend, node in nodes.get_all_nodes_with_backends()]
+    locationChoices.insert(0, ('-1', _("Don't mock location")))
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: This is ugly. Define outside function which returns iterator which provides all this data. So first it provides non-mocking value, then the rest.

Comment (mitar) [inline piplmesh/frontend/forms.py]: You don't have to use `-1` value there. It could be simply an empty string. :-) Because it is strange to have strings for other values, and then number for this value. Empty string is much easier to understand: nothing should be mocked.

File: piplmesh/nodes/__init__.py (hunk 21 above)
Comment (mitar) [inline piplmesh/nodes/__init__.py]: This if is not really correct. You probably want that if mocking is in effect (request.session[MOCKING_SESSION_KEY] is true) and `allow_mocking` is false, then this node value which was retrieved should not be allowed to be passed on! So you should set it to `None` in such case.

File: piplmesh/frontend/templates/header.html (hunk 22 above)
Comment (mitjap) [inline piplmesh/frontend/templates/header.html]: <li> tag (and some content that is hidden) is shown as you predicted. Will add a surrounding if around.

**Comment by mitar** [comment]
Don't forget to update from main repository.

File: piplmesh/frontend/forms.py [hunk 28]
```diff
@@ -2,6 +2,10 @@
 from django.forms import widgets
 from django.utils.translation import ugettext_lazy as _
 
+import itertools
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: Not needed.

File: piplmesh/frontend/forms.py [hunk 29]
```diff
@@ -10,3 +14,12 @@ class ContactForm(forms.Form):
     subject = forms.CharField(label=_("Subject"))
     email = forms.EmailField(label=_("Your e-mail address"))
     message = forms.CharField(widget=widgets.Textarea())
+
+def locationChoices():
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: We use _ naming for functions.

File: piplmesh/frontend/forms.py [hunk 30]
```diff
@@ -10,3 +14,12 @@ class ContactForm(forms.Form):
...
+    yield ('', _("Don't mock location"))
+    for backend, node in nodes.get_all_nodes_with_backends():
+        yield ("%s-%s" % (backend.get_full_name(), node.id), node.name)
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: For constant strings, we use `'`.

File: piplmesh/frontend/static/piplmesh/js/common_staff.js [hunk 31]
```diff
@@ -0,0 +1,7 @@
+$(document).ready(function () {
+    // Locations
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common_staff.js]: This comment is really unnecessary. It does not tell you anything new.

File: piplmesh/frontend/static/piplmesh/js/common_staff.js [hunk 32]
```diff
@@ -0,0 +1,7 @@
...
+    $('#id_locations').change(function (event) {
+        $(this).closest('form').submit();
+    });
+
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/js/common_staff.js]: Unnecessary line.

File: piplmesh/frontend/templates/header.html [hunk 33]
```diff
@@ -21,8 +21,17 @@
                 {% trans "See your query results on " %}<a id="google_results_link">Google</a>.
             {% endif %}
         </div>
+
         {% with next|default:request_get_next|default:request.get_full_path as next_url %}
             <ul class="user">
+                {% if locations_form %}
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: I would add here also `user.is_authenticated and user.is_staff` to the mix. Not that it is necessary, but to the reader tells easier when this thing goes through.

File: piplmesh/frontend/views.py [hunk 34]
```diff
@@ -64,6 +65,25 @@ class UserView(detail.DetailView):
...
+    # TODO: redirect to initiator page
+    success_url = urlresolvers.reverse_lazy('home')
+
+    def form_valid(self, form):
+        location = form.cleaned_data['locations']
+
+        if location == '':
+            nodes.flush_session(self.request)
+        else:
+            node_backend, node_id = location.split('-')
```
Comment (mitar) [inline piplmesh/frontend/views.py]: You have to limit how many times it splits.

File: piplmesh/frontend/views.py [hunk 35]
```diff
@@ -64,6 +65,25 @@ class UserView(detail.DetailView):
     slug_field = 'username'
     slug_url_kwarg = 'username'
 
+class LocationsView(generic_views.FormView):
+    form_class = forms.LocationsForm
+
+    # TODO: redirect to initiator page
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Start TODO comments with upper-case letter.

File: piplmesh/frontend/views.py [hunk 36]
```diff
@@ -64,6 +65,25 @@ class UserView(detail.DetailView):
     slug_field = 'username'
     slug_url_kwarg = 'username'
 
+class LocationsView(generic_views.FormView):
```
Comment (mitar) [inline piplmesh/frontend/views.py:68]: You are forgetting to validate if user is really authenticated and staff. Otherwise somebody can fake this request (especially because we are an open source project and can see how the request should look like).

File: piplmesh/nodes/__init__.py [hunk 37]
```diff
@@ -65,8 +66,16 @@ def get_node(request):
     try:
         node_id = request.session[SESSION_KEY]
         backend_path = request.session[BACKEND_SESSION_KEY]
+        mocking = request.session.get(MOCKING_SESSION_KEY, False)
```
Comment (mitar) [inline piplmesh/nodes/__init__.py]: Move this to the end of this variables. It is the least important, or it is a special value.

File: piplmesh/nodes/__init__.py [hunk 38]
```diff
@@ -65,8 +66,16 @@ def get_node(request):
...
         backend = load_backend(backend_path)
         node = backend.get_node(node_id)
+
```
Comment (mitar) [inline piplmesh/nodes/__init__.py:72]: Unnecessary empty line.

File: piplmesh/nodes/__init__.py [hunk 39]
```diff
@@ -65,8 +66,16 @@ def get_node(request):
...
+
+        if allow_mocking and mocking:
+            return node
```
Comment (mitar) [inline piplmesh/nodes/__init__.py:75]: I would again verify here that user is authenticated and that is staff.

File: piplmesh/nodes/__init__.py [hunk 40]
```diff
@@ -65,8 +66,16 @@ def get_node(request):
...
+        elif mocking:
+            node = None
```
Comment (mitar) [inline piplmesh/nodes/__init__.py:78]: A comment describing what are you doing here would be useful.

File: piplmesh/nodes/backends.py [hunk 41]
```diff
@@ -23,12 +23,12 @@ def get_closest_node(self, request, latitude, longitude):
 
         node_id = random.randrange(len(data.nodes))
         node = copy.copy(data.nodes[node_id])
-        node.id = node_id
+        node.id = str(node_id)
```
Comment (mitar) [inline piplmesh/nodes/backends.py]: No, there should be no need for this.

File: piplmesh/nodes/backends.py [hunk 42]
```diff
@@ -37,5 +37,8 @@ def get_node(self, node_id):
     def get_all_nodes(self):
         for i, node in enumerate(data.nodes):
             node = copy.copy(node)
-            node.id = i
+            node.id = str(i)
```
Comment (mitar) [inline piplmesh/nodes/backends.py]: No, there should be no need for this.

Comment (mitar) [inline piplmesh/nodes/backends.py]: Caller should convert to string when necessary. ID object should provide a way to convert it to string.

File: piplmesh/panels/map/static/piplmesh/panel/map/panel.js [hunk 43]
```diff
@@ -13,7 +13,7 @@ $(document).ready(function () {
         'mapTypeControlOptions': {
             'mapTypeIds': [
                 google.maps.MapTypeId.ROADMAP,
-                google.maps.MapTypeId.SATELLITE
+                google.maps.MapTypeId.SATELLITE,
```
Comment (mitar) [inline piplmesh/panels/map/static/piplmesh/panel/map/panel.js:18]: See, because you mixed other things in your pull request, you now have to resolve conflicts with main repository. Pull from it and merge.

(You can fix such things locally, but there is no need to commit them into this pull request.)

File: piplmesh/frontend/context_processors.py [hunk 44]
```diff
@@ -21,4 +20,6 @@ def global_vars(request):
         # Variables
         'logo_url': "piplmesh/images/logo-%s.png" % translation.get_language(),
         'request_get_next': request.REQUEST.get(auth.REDIRECT_FIELD_NAME),
-    }
+
+        # Forms
+        'locations_form' : forms.LocationsForm() if getattr(request, 'user', None) and request.user.is_authenticated() and request.user.is_staff else None,    }
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: `}` into new line.

File: piplmesh/frontend/views.py [hunk 45]
```diff
@@ -3,10 +3,12 @@
 from django import dispatch, http, template
 from django.conf import settings
 from django.contrib import messages
+from django.contrib.auth.decorators import login_required
```
Comment (mitar) [inline piplmesh/frontend/views.py]: You are not using this?

File: piplmesh/frontend/views.py [hunk 46]
```diff
@@ -3,10 +3,12 @@
...
 from django.core import mail, urlresolvers
 from django.core.files import storage
 from django.test import client
 from django.utils import simplejson
+from django.utils.decorators import method_decorator
```
Comment (mitar) [inline piplmesh/frontend/views.py]: You are not using this?

File: piplmesh/frontend/views.py [hunk 47]
```diff
@@ -64,6 +67,33 @@ class UserView(detail.DetailView):
...
 
+class LocationsView(generic_views.FormView):
+    form_class = forms.LocationsForm
+
+    # TODO: Redirect to initiator page
+    success_url = urlresolvers.reverse_lazy('home')
+
+    def form_valid(self, form):
+        user = self.request.user
+
```
Comment (mitar) [inline piplmesh/frontend/views.py]: What is this? Some stale code. Clean!

File: piplmesh/frontend/views.py [hunk 48]
```diff
@@ -64,6 +67,33 @@ class UserView(detail.DetailView):
...
+        else:
+            node_backend, node_id = location.split('-', 1)
+            self.request.session[nodes.SESSION_KEY] = node_id
+            self.request.session[nodes.BACKEND_SESSION_KEY] = node_backend
+            self.request.session[nodes.MOCKING_SESSION_KEY] = True
+
+        return super(LocationsView, self).form_valid(form)
+
+    def dispatch(self, request, *args, **kwargs):
+        staff = request.user
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Don't rename variables. And probably you do not need to store request.user into a variable. Just access it directly.

File: piplmesh/frontend/views.py [hunk 49]
```diff
@@ -64,6 +67,33 @@ class UserView(detail.DetailView):
...
+        if not (staff and staff.is_authenticated() and staff.is_staff):
```
Comment (mitar) [inline piplmesh/frontend/views.py]: In general, positive conditions is easier to think for humans. So:

```
if request.user and request.user.is_authenticated() and request.use.is_staff):
   return super..
raise PermissionDenied
```

File: piplmesh/nodes/__init__.py [hunk 50]
```diff
@@ -67,6 +68,16 @@ def get_node(request):
         backend_path = request.session[BACKEND_SESSION_KEY]
         backend = load_backend(backend_path)
         node = backend.get_node(node_id)
+        mocking = request.session.get(MOCKING_SESSION_KEY, False)
+
+        # Return node if mocking is in progress and user is authenticated
+        if allow_mocking and mocking and request.user and request.user.is_authenticated() and request.user.is_staff:
+            return node
+        # If mocking was in progress and user isn't allowed to mock reset nodes value.
```
Comment (mitar) [inline piplmesh/nodes/__init__.py]: No `.` at the end.

Comment (mitjap) [inline piplmesh/nodes/__init__.py]: that dot looked like stain on my screen :)

File: piplmesh/frontend/context_processors.py [hunk 51]
```diff
@@ -3,13 +3,12 @@
...
 
 from piplmesh import urls
-from piplmesh.frontend import tasks
+from piplmesh.frontend import forms, tasks
 
 def global_vars(request):
     """
     Adds global context variables to the context.
     """
-
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py:12]: Why removed this line?

File: piplmesh/frontend/forms.py [hunk 52]
```diff
@@ -10,3 +12,12 @@ class ContactForm(forms.Form):
...
     message = forms.CharField(widget=widgets.Textarea())
+
+def location_choices():
+    yield ('', _("Don't mock location"))
+    for backend, node in nodes.get_all_nodes_with_backends():
+        yield ('%s-%s' % (backend.get_full_name(), node.id), node.name)
+
+class LocationsForm(forms.Form):
+    locations = forms.ChoiceField(choices=location_choices(), label='', required=False)
+
```
Comment (mitar) [inline piplmesh/frontend/forms.py]: One line too much. :-)

File: piplmesh/frontend/views.py [hunk 53]
```diff
@@ -4,6 +4,7 @@
 from django.conf import settings
 from django.contrib import messages
 from django.core import mail, urlresolvers
+from django.core.exceptions import PermissionDenied
```
Comment (mitar) [inline piplmesh/frontend/views.py]: Import only modules. So import just exceptions. And then access through that.

**Comment by mitar** [comment]
Good. Otherwise looks good.
